File: src/main/scala/hohserg/endothermic/example/java/MutableImmutableExample.java
package hohserg.endothermic.example.java;

import com.google.common.collect.Lists;
import hohserg.endothermic.quad.BaseUnpackedQuad;
import hohserg.endothermic.quad.immutable.LazyUnpackedQuad;
import net.minecraft.client.renderer.block.model.BakedQuad;

import java.util.List;

public class MutableImmutableExample {

    public static List<BakedQuad> immutableVersion(BakedQuad quad) {
        LazyUnpackedQuad q1 = LazyUnpackedQuad.apply(quad);
        BaseUnpackedQuad translated = q1.translate(0.1f, 0, 1);
        BaseUnpackedQuad rotated = q1.rotate((float) Math.PI, 0, 1, 0);
        //q1 != translated; q1 != rotated ; translated != rotated
        return Lists.newArrayList(translated.toBakedQuad(), rotated.toBakedQuad());
    }

    public static List<BakedQuad> mutableVersion(BakedQuad quad) {
        hohserg.endothermic.quad.mutable.LazyUnpackedQuad q1 = hohserg.endothermic.quad.mutable.LazyUnpackedQuad.apply(quad);

        BaseUnpackedQuad translated = q1.copy().translate(0.1f, 0, 1);//copy for save original version
        // q1 != translated

        BaseUnpackedQuad rotated = q1.rotate((float) Math.PI, 0, 1, 0);//no copy, no new instance creating
        // q1 == rotated

        return Lists.newArrayList(translated.toBakedQuad(), rotated.toBakedQuad());
    }
}

File: src/main/scala/hohserg/endothermic/example/scala/MutableImmutableExample.scala
package hohserg.endothermic.example.scala

import java.util

import com.google.common.collect.Lists
import hohserg.endothermic.quad.{immutable, mutable}
import net.minecraft.client.renderer.block.model.BakedQuad

class MutableImmutableExample {
  def immutableVersion(quad: BakedQuad): util.List[BakedQuad] = {
    val q1 = immutable.LazyUnpackedQuad.apply(quad)
    val translated = q1.translate(0.1f, 0, 1)
    val rotated = q1.rotate(Math.PI.toFloat, 0, 1, 0)
    //q1 != translated; q1 != rotated ; translated != rotated
    Lists.newArrayList(translated.toBakedQuad, rotated.toBakedQuad)
  }

  def mutableVersion(quad: BakedQuad): util.List[BakedQuad] = {
    val q1 = mutable.LazyUnpackedQuad.apply(quad)
    val translated = q1.copy.translate(0.1f, 0, 1) //copy for save original version
    // q1 != translated

    val rotated = q1.rotate(Math.PI.toFloat, 0, 1, 0) //no copy, no new instance creating
    // q1 == rotated
    Lists.newArrayList(translated.toBakedQuad, rotated.toBakedQuad)
  }
}
